include/zarray/zarray_buffer_handler.hpp

```diff
@@ -0,0 +1,85 @@
+/***************************************************************************
+* Copyright (c) Johan Mabille, Sylvain Corlay and Wolf Vollprecht          *
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_ZARRAY_BUFFER_HANDLER_HPP
+#define XTENSOR_ZARRAY_BUFFER_HANDLER_HPP
+
+#include <array>
+#include <set>
+#include <map>
+#include <memory>
+
+namespace xt
+{
+    // forward declare
+    class zarray_impl;
+
+    namespace detail
+    {
+        class zarray_buffer_handler
+        {
+        public:
+
+            zarray_buffer_handler(zarray_impl & res)
+            :   m_res(res),
+                m_buffers(),
+                m_free_buffers()
+            {
+                this->mark_as_free(&res);
+            }
+
+            auto get_free_buffer(const std::size_t type_index)
```

Comment:
I am a bit afraid of assignments to integers. 
xtensor and zarray would then give differnt results.
```c++
xarray<float> x0 = xt::ones<float>({2,2}) * 0.1f; 
xarray<float> x1 = xt::ones<float>({2,2}) * 0.1f;
xarray<int> xres = xarray<int>::from_shape({2,2});

zarray z0(x0);
zarray z1(x1);

// this would fill xres with 0
zarray zres = 10 * (z0 + z1);

// this would fill xres with 2
xres = 10 * (z0 + z1);

```

Reply:
Oh good remark. So no choice, we need to keep intermediate result type.